### PR TITLE
Disable DiagnosticTrustManager in FIPS 140

### PR DESCRIPTION
**Pull request by jkakavas**
This commit changes the default behavior for 
`xpack.security.ssl.diagnose.trust` when running in a FIPS 140 JVM.

More specifically, when `xpack.security.fips_mode.enabled` is true:
- If `xpack.security.ssl.diagnose.trust` is not explicitly set, the 
default value of it becomes `false` and a log message is printed
on info level, notifying of the fact that the TLS/SSL diagnostic 
messages are not enabled when in a FIPS 140 JVM. 
- If `xpack.security.ssl.diagnose.trust` is explicitly set, the value of
it is honored, even in FIPS mode. 

This is relevant only for 7.x where we support Java 8 in which 
SunJSSE can still be used as a FIPS 140 provider for TLS. SunJSSE
in FIPS mode, disallows the use of other TrustManager implementations
than the one shipped with SunJSSE.

### CODE REVIEW

**Comment by tvernum** [comment]
> If xpack.security.ssl.diagnose.trust is explicitly set to true, we fail

Is that what we want? Do we support using BC-FIPS-JSSE on 7.x? (I think we do/will) 
If so, isn't it OK for users to opt-in to diagnostic trust if their JSSE provider supports it?

**Comment by jkakavas** [comment]
> Is that what we want? Do we support using BC-FIPS-JSSE on 7.x? (I think we do/will)
If so, isn't it OK for users to opt-in to diagnostic trust if their JSSE provider supports it?

You're right, I was probably too aggressively defensive here. We do plan to start testing with BCJSSE in  FIPS mode and there is no reason to not support this.

File: docs/reference/settings/security-settings.asciidoc [hunk 1]
```diff
@@ -1505,6 +1505,8 @@ to establish trust.
 This diagnostic message contains information that can be used to determine the
 cause of the failure and assist with resolving the problem.
 Set to `false` to disable these messages.
+NOTE: This defaults to `false` and cannot be enabled when
+`xpack.security.fips_mode.enabled` is `true`.
```
Comment (tvernum) [inline docs/reference/settings/security-settings.asciidoc]: Based on agreement to allow FIPS users to opt-in, let's remove the 2nd part of this note.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java [hunk 2]
```diff
@@ -844,4 +837,13 @@ private static String sslContextAlgorithm(List<String> supportedProtocols) {
...
             + supportedProtocols);
     }
+
+    private boolean shouldEnableDiagnoseTrust() {
+        if (XPackSettings.FIPS_MODE_ENABLED.get(settings)) {
+            logger.info("diagnostic messages for SSL/TLS trust failures are not enabled in FIPS 140 mode.");
+            return false;
+        } else {
+            return DIAGNOSE_TRUST_EXCEPTIONS_SETTING.get(settings);
+        }
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java]:
```
       if (XPackSettings.FIPS_MODE_ENABLED.get(settings) && DIAGNOSE_TRUST_EXCEPTIONS_SETTING.exists(settings) == false ) {
            logger.info("diagnostic messages for SSL/TLS trust failures are not enabled in FIPS 140 mode.");
            return false;
        } else {
            return DIAGNOSE_TRUST_EXCEPTIONS_SETTING.get(settings);
        }
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 3]
```diff
@@ -903,7 +903,11 @@ static void validateForFips(Settings settings) {
             validationErrors.add("Only PBKDF2 is allowed for password hashing in a FIPS 140 JVM. Please set the " +
                 "appropriate value for [ " + XPackSettings.PASSWORD_HASHING_ALGORITHM.getKey() + " ] setting.");
         }
-
+        if (XPackSettings.DIAGNOSE_TRUST_EXCEPTIONS_SETTING.exists(settings)
+            && XPackSettings.DIAGNOSE_TRUST_EXCEPTIONS_SETTING.get(settings)) {
+            validationErrors.add("SSL/TLS diagnostic messages cannot be enabled in a FIPS 140 JVM. Please set [ " +
+                XPackSettings.DIAGNOSE_TRUST_EXCEPTIONS_SETTING.getKey() + " ] to false or remove this setting from your configuration. ");
+        }
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: I think we can just drop this.

**Review by tvernum — CHANGES_REQUESTED** [review]
I've marked this as "request changes" so it drops from my queue (based on agreement above)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java [hunk 4]
```diff
@@ -852,4 +845,13 @@ private static String sslContextAlgorithm(List<String> supportedProtocols) {
         throw new IllegalArgumentException("no supported SSL/TLS protocol was found in the configured supported protocols: "
             + supportedProtocols);
     }
+
+    private boolean shouldEnableDiagnoseTrust() {
+        if (XPackSettings.FIPS_MODE_ENABLED.get(settings) && DIAGNOSE_TRUST_EXCEPTIONS_SETTING.exists(settings) == false ) {
+            logger.info("diagnostic messages for SSL/TLS trust failures are not enabled in FIPS 140 mode be default.");
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java]:
```suggestion
            logger.info("diagnostic messages for SSL/TLS trust failures are not enabled in FIPS 140 mode by default.");
```

**Review by tvernum — APPROVED** [review]

**Pull request merged**